### PR TITLE
[feat]curveadm: add export cmd for export the yaml that tools-v2 used

### PR DESCRIPTION
**Pull request by youarefree123**
Fixed: https://github.com/opencurve/curveadm/issues/315

add export cmd for export the yaml that tools-v2 used

### CODE REVIEW

File: internal/errno/errno.go [hunk 1]
```diff
@@ -280,6 +280,9 @@ var (
 	// 222: command options (client/fs)
 	ERR_FS_MOUNTPOINT_REQUIRE_ABSOLUTE_PATH = EC(222000, "mount point must be an absolute path")
 
+	// 223: command options (export)
+	ERR_EXPORT_TOOLSV2_CONF_REQUIRE_ABSOLUTE_PATH = EC(222000, "/path/to/curve.yaml must be an absolute path")
```
Comment (montaguelhz) [inline internal/errno/errno.go:284]: maybe it is 223000

File: cli/command/export.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
```
Comment (montaguelhz) [inline cli/command/export.go:2]: 2023

File: cli/command/export.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	// 2) generate get export playbook
+	pb, err := genExportPlaybook(curveadm, dcs, options)
+	if err != nil {
+		return err
+	}
+
+	// 3) run playground
+	err = pb.Run()
+
+	// 4) print success prompt
```
Comment (montaguelhz) [inline cli/command/export.go:129]: The comment should come after error handling

File: internal/task/task/common/export_toolsv2_conf.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
```
Comment (montaguelhz) [inline internal/task/task/common/export_toolsv2_conf.go:2]: ditto

File: internal/task/task/common/sync_config.go [hunk 5]
```diff
@@ -91,11 +91,11 @@ func newToolV2Mutate(dc *topology.DeployConfig, delimiter string, forceRender bo
 
 		// replace variable
 		value, err = dc.GetVariables().Rendering(value)
-		if err != nil {
-			return
+		if err == nil {
+			out = fmt.Sprintf("%s%s %s", key, delimiter, value)
+		} else { // FIX : add replace warring log
+			err = nil
```
Comment (montaguelhz) [inline internal/task/task/common/sync_config.go:97]: Why make the error nil？

Comment (youarefree123) [reply internal/task/task/common/sync_config.go:97]: When I don't have a snapshot service configured, the Rendering function will introduce an error as shown in the following：
image

I am aware that the current approach to handling this issue is not particularly elegant. Do you have any valuable suggestions for improvement?

Comment (montaguelhz) [reply internal/task/task/common/sync_config.go:97]: Maybe you can add a check before it, or talk it with your mentor or  @Cyber-SiKu for a better  approach.

File: cli/command/export.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/opencurve/curveadm/internal/configure/topology"
+	"github.com/opencurve/curveadm/internal/errno"
+	"github.com/opencurve/curveadm/internal/playbook"
+	cliutil "github.com/opencurve/curveadm/internal/utils"
+	"github.com/spf13/cobra"
+)
+
+var (
+	GET_EXPORT_PLAYBOOK_STEPS = []int{
+		playbook.SYNC_CONFIG,
```
Comment (caoxianfei1) [inline cli/command/export.go]: Why need the step?

File: cli/command/export.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		playbook.EXPORT_TOOLSV2_CONF,
+	}
+)
+
+type exportOptions struct {
+	output string
```
Comment (caoxianfei1) [inline cli/command/export.go:48]: --path is better?

Comment (caoxianfei1) [inline cli/command/export.go:48]: which one host to export? Specify host?

File: cli/command/export.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+				},
+			})
+		} else {
+			pb.AddStep(&playbook.PlaybookStep{
+				Type:    step,
+				Configs: dcs,
+				Options: map[string]interface{}{
+					comm.KEY_TOOLSV2_CONF_PATH: options.output,
+				},
+			})
```
Comment (caoxianfei1) [inline cli/command/export.go]: Ditto.

File: internal/task/task/common/export_toolsv2_conf.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	containerId, err := curveadm.Storage().GetContainerId(serviceId)
+	if curveadm.IsSkip(dc) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	} else if len(containerId) == 0 {
+		return nil, nil
+	} else if containerId == comm.CLEANED_CONTAINER_ID {
+		return nil, nil
+	}
```
Comment (caoxianfei1) [inline internal/task/task/common/export_toolsv2_conf.go]: What does mean?

File: internal/task/task/common/sync_config.go (hunk 5 above)
Comment (caoxianfei1) [inline internal/task/task/common/sync_config.go]: Ditto.

**Comment by caoxianfei1** [comment]
BTW, please push it to develop branch again.

File: cli/command/export.go (hunk 6 above)
Comment (youarefree123) [reply cli/command/export.go]: Originally, I used it on the cluster I had started, and the conf in the container did not render successfully, so I added it, but now it seems that this step is really not needed, and I will restart this cluster later, and then test it.

File: cli/command/export.go (hunk 7 above)
Comment (youarefree123) [reply cli/command/export.go:48]: Okay, I'll make changes

Comment (youarefree123) [reply cli/command/export.go:48]: I found that all hosts include curve.yaml. I think the conf in each host should be the same, so currently I have selected dcs[0], just like the implementation in the “status” command.
[status.go](https://github.com/opencurve/curveadm/blob/580b26dd06144f327a9afe5e103bd18ff7a8a1a9/cli/command/status.go#L118-L120)

File: internal/task/task/common/export_toolsv2_conf.go (hunk 9 above)
Comment (youarefree123) [reply internal/task/task/common/export_toolsv2_conf.go]: I referred to the implementation of [NewCollectService](https://github.com/opencurve/curveadm/blob/580b26dd06144f327a9afe5e103bd18ff7a8a1a9/internal/task/task/common/collect_service.go#L79-L87) at that time, may be I need to evaluate whether it is necessary.

File: internal/task/task/common/sync_config.go (hunk 5 above)
Comment (youarefree123) [reply internal/task/task/common/sync_config.go]: Please refer to the discussion [here.](https://github.com/opencurve/curveadm/pull/344#discussion_r1393567785) Perhaps I need to modify the logic of sync_config? Or is there any other more suitable method?

Comment (caoxianfei1) [reply internal/task/task/common/sync_config.go]: Yes, if render failed, you can open another pull request to fix it.

**Pull request closed without merging**